provider/ovh.go

```diff
@@ -0,0 +1,354 @@
+	ErrNoDryRun = errors.New("dry run not supported")
+)
+
+// OVHProvider is an implementation of Provider for OVH DNS.
+type OVHProvider struct {
+	client ovhClient
+
+	domainFilter DomainFilter
+	DryRun       bool
+}
+
+type ovhClient interface {
+	Post(string, interface{}, interface{}) error
+	Get(string, interface{}) error
+	Delete(string, interface{}) error
+}
+
+type ovhRecordFields struct {
+	FieldType string `json:"fieldType"`
+	SubDomain string `json:"subDomain"`
+	TTL       int64  `json:"ttl"`
+	Target    string `json:"target"`
+}
+
+type ovhRecord struct {
+	ovhRecordFields
+	ID   uint64 `json:"id"`
+	Zone string `json:"zone"`
+}
+
+type ovhChange struct {
+	ovhRecord
+	Action int
+}
+
+// NewOVHProvider initializes a new OVH DNS based Provider.
+func NewOVHProvider(ctx context.Context, domainFilter DomainFilter, endpoint string, generateConsumerKey bool, dryRun bool) (*OVHProvider, error) {
+	client, err := ovh.NewEndpointClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	if generateConsumerKey {
+		ckReq := client.NewCkRequest()
+		ckReq.AddRules(ovh.ReadOnly, "/domain/zone")
+		ckReq.AddRecursiveRules(ovh.ReadWrite, "/domain/zone/*/record")
+		ckReq.AddRules([]string{"POST"}, "/domain/zone/*/refresh")
+		response, err := ckReq.Do()
+		if err != nil {
+			return nil, err
+		}
+		log.Infof("Generated consumer key: %s\n", response.ConsumerKey)
+		log.Infof("Please visit %s to validate it\n", response.ValidationURL)
+		return nil, fmt.Errorf("You have to validated the consumer key")
+	}
+	// TODO: Add Dry Run support
+	if dryRun {
+		return nil, ErrNoDryRun
+	}
+	return &OVHProvider{
+		client:       client,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}, nil
+}
+
+// Records returns the list of records in all relevant zones.
+func (p *OVHProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+
+	_, records, err := p.zonesRecords(ctx)
+	if err != nil {
+		return nil, err
+	}
+	endpoints := ovhGroupByNameAndType(records)
+	log.Infof("OVH: %d endpoints have been found\n", len(endpoints))
+	return endpoints, nil
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *OVHProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	zones, records, err := p.zonesRecords(ctx)
+	zonesChangeUniques := map[string]bool{}
+	if err != nil {
+		return err
+	}
+
+	allChanges := make([]ovhChange, 0, countTargets(changes.Create, changes.UpdateNew, changes.UpdateOld, changes.Delete))
+
+	allChanges = append(allChanges, newOvhChange(ovhCreate, changes.Create, zones, records)...)
+	allChanges = append(allChanges, newOvhChange(ovhCreate, changes.UpdateNew, zones, records)...)
+
+	allChanges = append(allChanges, newOvhChange(ovhDelete, changes.UpdateOld, zones, records)...)
+	allChanges = append(allChanges, newOvhChange(ovhDelete, changes.Delete, zones, records)...)
+
+	log.Infof("OVH: %d changes will be done\n", len(allChanges))
+
+	eg, _ := errgroup.WithContext(ctx)
```

Comment:
Always check for errors.

Reply:
It's a litte misleading, but the second argument is not an error (But the created context)
See https://github.com/golang/sync/blob/master/errgroup/errgroup.go#L32

Reply:
👍